### PR TITLE
Feature/query microservices

### PR DESCRIPTION
**Pull request by jwomeara**

### CODE REVIEW

File: src/main/java/datawave/microservice/query/storage/CachedQueryStatus.java
Comment (jwomeara) [inline src/main/java/datawave/microservice/query/storage/CachedQueryStatus.java:1]: I have a variety of concerns with this class, mainly over how often we will be hitting the cache.  

Instead of periodically updating the QueryStatus under the covers (which could result in any number of unnecessary updates being performed) I think we should be more deliberate when it comes to hitting up the cache for an update.  If we don't need one, we shouldn't be doing updates in the background.

I am also concerned about the idea of certain setters triggering a forced cache update.  This is a situation where I feel like hiding this functionality in the setter could lead to unintended updates.  There is nothing to indicate to the developer that a method is going to force a cache update.  My concern is that this could cause us to update the cache prematurely/unintentionally in the future.  

Another concern I have is how this class both IS a QueryStatus, but also delegates to an internal QueryStatus.  The issue here is that as we add new parameters to the QueryStatus object, we will also have to add equivalent methods in this class, or else we will be updating the actual QueryStatus, and not the delegate QueryStatus, causing us to potentially miss updates.  This adds additional upkeep when making a change, seems highly error prone (and similar to the relationship between a QueryLogic and QueryConfiguration which I thought we were trying to get away from), and we have already violated this contract by adding parameters to QueryStatus which do not have delegate methods in CachedQueryStatus.  Thus, updates to those parameters will be lost unless we add delegating methods.  

Another aspect of this class that just feels a little clunky to me is this idea of using a Runnable (which implies running code in a separate thread) simply for it's ability to act as a lambda.  In my opinion a Supplier would be better for this use case.

File: src/main/java/datawave/microservice/query/mapreduce/jobs/BulkResultsJob.java [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+        systemProperties.put(MapReduceQueryProperties.PREFIX + ".enabled", "false");
+        
+        // TODO: JWO: Set properties for the data and edge dictionaries if necessary
+        
+        // TODO: JWO: Set properties for the user's JWT
+        
+        // TODO: JWO: What is the substitute for this?
+        // if (this.jsseSecurityDomain != null) {
+        // String useJobCacheString = systemProperties.getProperty("dw.mapreduce.securitydomain.useJobCache");
+        // boolean useJobCache = Boolean.parseBoolean(useJobCacheString);
```
[the remark this answers is not in the document]
Comment (jwomeara) [reply src/main/java/datawave/microservice/query/mapreduce/jobs/BulkResultsJob.java]: nope.  that looks like wildfly configuration stuff (as evidenceded by the 'dw.*' properties).  we shouldn't need that anymore.

File: src/main/java/datawave/microservice/query/storage/QueryStatus.java [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+import datawave.core.query.configuration.GenericQueryConfiguration;
+import datawave.core.query.logic.QueryKey;
+import datawave.microservice.authorization.user.DatawaveUserDetails;
+import datawave.microservice.query.Query;
+import datawave.microservice.querymetric.BaseQueryMetric.Prediction;
+import datawave.webservice.query.exception.DatawaveErrorCode;
+
+public class QueryStatus implements Serializable {
+    /**
+     * These are the possible query states correlating with the activity that the user is requesting. DEFINE: define a query CREATE: create a query and get
```
Comment (apmoriarty) [inline src/main/java/datawave/microservice/query/storage/QueryStatus.java:31]: small nit, the docs in class won't render in a list

Comment (jwomeara) [reply src/main/java/datawave/microservice/query/storage/QueryStatus.java:31]: sorry, i don't follow.  what do you mean?

File: src/main/java/datawave/microservice/query/logic/config/converter/StringToTypeConverter.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.stereotype.Component;
+
+import datawave.data.type.Type;
+
+@Component
+@ConfigurationPropertiesBinding
+public class StringToTypeConverter implements Converter<String,Type> {
+    @Override
+    public Type convert(String source) {
```
Comment (alerman) [inline src/main/java/datawave/microservice/query/logic/config/converter/StringToTypeConverter.java:13]: Is there not method like this in a library? Not necessarily saying we use it, just seems like it would be code that is duplicated. Maybe thats ok.

File: src/main/java/datawave/microservice/query/storage/QueryStatus.java (hunk 2 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/query/storage/QueryStatus.java:31]: i see what you mean.  working it

File: src/main/java/datawave/microservice/query/logic/config/converter/StringToTypeConverter.java (hunk 3 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/query/logic/config/converter/StringToTypeConverter.java:13]: Hmmm, i'm not sure.  This is specifically used to map yaml spring properties to actual datawave Types, so im not sure that's something we've done before.  Looking in the type-utils project, I'm not seeing anything at first glance that could serve this purpose.

File: src/main/java/datawave/microservice/query/storage/QueryStatus.java (hunk 2 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/query/storage/QueryStatus.java]: should be good now.

File: src/main/java/datawave/microservice/query/storage/CachedQueryStatus.java [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+        QueryStorageLock lock = cache.getQueryStatusLock(queryId);
+        lock.lock();
+        try {
+            forceCacheUpdateInsideSetter();
+            updater.run();
+            cache.updateQueryStatus(queryStatus);
+        } finally {
+            lock.unlock();
+        }
+    }
```
Comment (apmoriarty) [inline src/main/java/datawave/microservice/query/storage/CachedQueryStatus.java:56]: Is there a use case for a bulk update method? I could see a scenario where multiple updates are made one after the other, and that would cause repeated locking and unlocking. 

If the situation I'm describing wouldn't arise, please ignore me.

File: src/main/java/datawave/microservice/query/messaging/kafka/KafkaQueryResultsManager.java [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    public void emptyQuery(String name) {
+        TopicDescription topic = describeTopic(name);
+        if (topic != null) {
+            Map<TopicPartition,RecordsToDelete> partitions = new HashMap<>();
+            RecordsToDelete records = RecordsToDelete.beforeOffset(Long.MAX_VALUE);
+            for (TopicPartitionInfo info : topic.partitions()) {
+                TopicPartition partition = new TopicPartition(name, info.partition());
+                partitions.put(partition, records);
+            }
+            DeleteRecordsResult result2 = adminClient.deleteRecords(partitions);
```
Comment (alerman) [inline src/main/java/datawave/microservice/query/messaging/kafka/KafkaQueryResultsManager.java:137]: Why result2 as opposed to result? If not using result consider a different more descriptive name

File: src/main/java/datawave/microservice/query/messaging/kafka/KafkaQueryResultsManager.java [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    /**
+     * Create a listener for a specified listener id
+     *
+     * @param listenerId
+     *            The listener id
+     * @return a query queue listener
+     */
+    @Override
+    public QueryResultsListener createListener(String listenerId, String queryId) {
+        createTopic(queryId);
```
Comment (alerman) [inline src/main/java/datawave/microservice/query/messaging/kafka/KafkaQueryResultsManager.java:61]: In Rabbit we use a QUERY_QUEUE_PREFIX. Should we also use a QUERY_TOPIC_PREFIX here?

File: src/main/java/datawave/microservice/query/messaging/rabbitmq/RabbitMQQueryResultsPublisher.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+import datawave.microservice.query.messaging.ClaimCheck;
+import datawave.microservice.query.messaging.QueryResultsPublisher;
+import datawave.microservice.query.messaging.Result;
+import datawave.microservice.query.messaging.config.MessagingProperties;
+
+class RabbitMQQueryResultsPublisher implements QueryResultsPublisher {
+    private final Logger log = LoggerFactory.getLogger(this.getClass());
+    
+    private static final long DEFAULT_MAX_MSG_SIZE = 536870912L;
```
Comment (alerman) [inline src/main/java/datawave/microservice/query/messaging/rabbitmq/RabbitMQQueryResultsPublisher.java:28]: Is this a Rabbit default? If not where did we come up with this constant? Consider adding a comment with why this number was chosen

File: src/main/java/datawave/microservice/query/storage/CachedQueryStatus.java
Comment (alerman) [reply src/main/java/datawave/microservice/query/storage/CachedQueryStatus.java:1]: My larger conern is the one @jwomeara mentioned re: this being both a QueryStatus and a delegate for QueryStatus. That feels off to me. In other similar use cases, I would expect that cache to provide a get/update method. I need to look more into how this is used, but i dont find this intuitive at the very least

File: src/main/java/datawave/microservice/query/storage/TaskStates.java [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.apache.commons.lang3.builder.ToStringBuilder;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.zaxxer.sparsebits.SparseBitSet;
```
Comment (alerman) [inline src/main/java/datawave/microservice/query/storage/TaskStates.java:15]: Do we need the memory optimization that using a sparsebitset provides over a hashset or other structure? I do see the memory benefits, but will the memory optimization be so great that it accounts for this class being harder to grok for someone not familiar?

File: src/main/java/datawave/microservice/query/storage/CachedQueryStatus.java
Comment (apmoriarty) [reply src/main/java/datawave/microservice/query/storage/CachedQueryStatus.java:1]: Mixing a delegate pattern while extending the delegate class is definitely off. It's also a violation of the principle of least responsibility -- if possible each class should do a minimum set of operations. The inconsistency of which methods will acquire and release a lock feels wrong. The client should know which methods are going to trigger lock acquisitions, and if the client is going to do multiple operations those should be batched with a single lock operation. 

In summary, I don't know why we need this. If there is indeed a problem that this code solves, it could be solved in a better way.

File: pom.xml [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+        <dependencies>
+            <dependency>
+                <groupId>com.google.guava</groupId>
+                <artifactId>guava</artifactId>
+                <version>${version.guava}</version>
+            </dependency>
+            <dependency>
+                <groupId>com.zaxxer</groupId>
+                <artifactId>SparseBitSet</artifactId>
+                <version>1.2</version>
```
Comment (ivakegg) [inline pom.xml:44]: move to a version constant

File: src/test/resources/config/application.yml [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+---
+
+spring.config.activate.on-profile: use-hazelcast
+
+datawave:
+  query:
+    messaging:
+      backend: hazelcast
+      hazelcast:
+        splitBrainProtectionEnabled: false
```
Comment (ivakegg) [inline src/test/resources/config/application.yml:93]: this feels dangerous

File: src/main/java/datawave/microservice/query/storage/CachedQueryStatus.java (hunk 4 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/query/storage/CachedQueryStatus.java:56]: What would a bulk update look like to you?  The way I understand it, the usage would be to pull the query status from the cache, make all of the updates you need to make to that local copy, and then as a locked operation push that back out to the cache to update the status.  For certain variables stored in the status, they might be updated by multiple services at the same time, and in that case we should lock the query id, pull the status, update the status (quickly), and then push it back out and unlock to ensure that the operation is atomic.  

I guess I need to understand more about the scenario you're thinking about.

File: src/main/java/datawave/microservice/query/messaging/kafka/KafkaQueryResultsManager.java (hunk 5 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/query/messaging/kafka/KafkaQueryResultsManager.java:137]: it might have made sense at one point, but it certainly doesn't now.  fixing.

File: src/main/java/datawave/microservice/query/messaging/kafka/KafkaQueryResultsManager.java (hunk 6 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/query/messaging/kafka/KafkaQueryResultsManager.java:61]: I think the intent was for both the queues and topics to just be named after the query id.  let me take a look at this and get back to you.

File: src/main/java/datawave/microservice/query/messaging/rabbitmq/RabbitMQQueryResultsPublisher.java (hunk 7 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/query/messaging/rabbitmq/RabbitMQQueryResultsPublisher.java:28]: I'll add a comment, but here's the reference:
https://github.com/rabbitmq/rabbitmq-common/blob/v3.8.0/include/rabbit.hrl#L238

File: src/main/java/datawave/microservice/query/storage/TaskStates.java (hunk 8 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/query/storage/TaskStates.java:15]: this worries me as well

File: src/test/resources/config/application.yml (hunk 10 above)
Comment (jwomeara) [reply src/test/resources/config/application.yml:93]: disabling split brain for tests seems dangerous to you?

File: src/main/java/datawave/microservice/query/messaging/kafka/KafkaQueryResultsManager.java (hunk 6 above)
Comment (ivakegg) [reply src/main/java/datawave/microservice/query/messaging/kafka/KafkaQueryResultsManager.java:61]: That was the idea, using the query id in both names.

File: src/main/java/datawave/microservice/query/storage/TaskStates.java (hunk 8 above)
Comment (ivakegg) [reply src/main/java/datawave/microservice/query/storage/TaskStates.java:15]: This was also for performance, and to minimize storage.  However I am happy to attempt to replace with something else.

File: src/test/resources/config/application.yml (hunk 10 above)
Comment (ivakegg) [reply src/test/resources/config/application.yml:93]: ah, ok.

File: src/main/java/datawave/microservice/query/storage/TaskStates.java (hunk 8 above)
Comment (ivakegg) [reply src/main/java/datawave/microservice/query/storage/TaskStates.java:15]: Replacing with a sorted set of integers...

File: src/main/java/datawave/microservice/query/messaging/kafka/KafkaQueryResultsManager.java (hunk 6 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/query/messaging/kafka/KafkaQueryResultsManager.java]: ok, i updated all of the results messaging subsystems to use the prefix.  all tested and working.

File: src/main/java/datawave/microservice/query/storage/CachedQueryStatus.java (hunk 4 above)
Comment (apmoriarty) [reply src/main/java/datawave/microservice/query/storage/CachedQueryStatus.java:56]: Based on my read of the setter methods that utilize the updateQueryStatus method, I could see a situation where the client is trying to update multiple setters at once within the context of processing a result page.
```java
// new page returned
cachedStatus.setNumResultsReturned()
cachedStatus.setLastPageNumber()
cachedStatus.setLastUsedMillis()
```
Each of those calls will trigger a lock/update/unlock cycle. It would make more sense to me if the flow was something like this
```java
// new page returned

// set the increment values without touching the cache
cachedStatus.setNumResultsReturned()
cachedStatus.setLastPageNumber()
cachedStatus.setLastUsedMillis()

cachedStatus.forceCacheUpdate()  // single lock, multiple variables updated
```

File: src/main/java/datawave/microservice/query/storage/TaskStates.java (hunk 8 above)
Comment (ivakegg) [reply src/main/java/datawave/microservice/query/storage/TaskStates.java]: done

File: src/main/java/datawave/microservice/query/storage/CachedQueryStatus.java (hunk 4 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/query/storage/CachedQueryStatus.java:56]: Ahh ok, I see.  yeah, in my opinion it should be
 - lock the cache for that id
 - get the latest status from the cache
 - update the values for those three things in the cache
 - push the status back into the cache
 - unlock the cache

i created a helper class which lets you pass a lambda to update the values in the status in a locked fashion.  i would prefer to see that used instead of this class.  

The class/method is here: 
https://github.com/NationalSecurityAgency/datawave-spring-boot-starter-query/blob/feature/queryMicroservices/src/main/java/datawave/microservice/query/util/QueryStatusUpdateUtil.java#L60

File: src/main/java/datawave/microservice/query/mapreduce/jobs/MapReduceJob.java [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+            }
+        }
+        
+        parameters.remove(AuditParameters.QUERY_SECURITY_MARKING_COLVIZ);
+        parameters.remove(AuditParameters.USER_DN);
+        parameters.remove(AuditParameters.QUERY_AUDIT_TYPE);
+    }
+    
+    /**
+     * @return HDFS FileSystem
```
Comment (cogross) [inline src/main/java/datawave/microservice/query/mapreduce/jobs/MapReduceJob.java:68]: give a description & @param for the javadoc

File: src/main/java/datawave/microservice/query/mapreduce/jobs/MapReduceJob.java [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+     * @throws IOException
+     */
+    protected final FileSystem getFileSystem(Configuration conf) throws IOException {
+        return FileSystem.get(conf);
+    }
+    
+    /**
+     * Common code to setup distributed cache and classpath for the job
+     *
+     * @param job
```
Comment (cogross) [inline src/main/java/datawave/microservice/query/mapreduce/jobs/MapReduceJob.java:78]: add @param baseDir, descriptions for the params & throws

Comment (jwomeara) [reply src/main/java/datawave/microservice/query/mapreduce/jobs/MapReduceJob.java:78]: sure.

File: src/main/java/datawave/microservice/query/storage/QueryStatus.java [hunk 13]
```diff
@@ -0,0 +1,443 @@
...
+        DEFINE, CREATE, PLAN, PREDICT, CLOSE, CANCEL, FAIL
+    }
+    
+    /**
+     * These are the possible stages of a query create/next/close<br>
+     * CREATE: Create the query.<br>
+     * PLAN: The query is being planned<br>
+     * TASK: planning is complete and next tasks are being generated<br>
+     * RESULTS: All next tasks have been generated and we are only generating results<br>
+     * DONE: The query id done (QUERY_STATE should be set to CLOSE, CANCEL, or FAIL).
```
Comment (cogross) [inline src/main/java/datawave/microservice/query/storage/QueryStatus.java:50]: No DONE in enum.

File: src/main/java/datawave/microservice/query/storage/CachedQueryStatus.java
Comment (ivakegg) [reply src/main/java/datawave/microservice/query/storage/CachedQueryStatus.java:1]: Removing the CachedQueryStatus mechanism and calling explicite cache.updateQueryStatus where needed.

File: src/main/java/datawave/microservice/query/storage/QueryStorageCacheImpl.java [hunk 14]
```diff
@@ -0,0 +1,493 @@
...
+    }
+    
+    /**
+     * Get a task for a given task key and lock it for processing. This return null if the task no longer exists. This will throw an exception if the task is
+     * already locked.
+     *
+     * @param taskKey
+     *            The task key
+     * @return The query task, null if deleted
+     * @throws TaskLockException
```
Comment (cogross) [inline src/main/java/datawave/microservice/query/storage/QueryStorageCacheImpl.java:402]: no exceptions thrown here, remove the annotation

File: src/main/java/datawave/microservice/query/storage/TaskStates.java [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+import datawave.util.StringUtils;
+
+public class TaskStates implements Serializable {
+    /**
+     * The possible task states: <br>
+     * READY: ready to run <br>
+     * RUNNING: currently running <br>
+     * COMPLETED: completed successfully <br>
+     * FAILED: failed to execute successfully <br>
+     * ORPHANED: orphaned after the query is closed or cancelled.
```
Comment (cogross) [inline src/main/java/datawave/microservice/query/storage/TaskStates.java:32]: ORPHANED not in the enum

File: src/main/java/datawave/microservice/query/storage/QueryStatus.java (hunk 13 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/query/storage/QueryStatus.java:50]: got it

File: src/main/java/datawave/microservice/query/storage/TaskStates.java (hunk 15 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/query/storage/TaskStates.java:32]: got it.  thanks

File: src/main/java/datawave/microservice/query/storage/QueryStorageCacheImpl.java (hunk 14 above)
Comment (jwomeara) [reply src/main/java/datawave/microservice/query/storage/QueryStorageCacheImpl.java:402]: good catch.  i didn't see this exception being thrown anywhere actually, so i deleted it.

**Review by avgAGB — APPROVED** [review]
